### ISSUE TITLE
Export Slots On Octopus Go 

### ISSUE THREAD

**Issue by nbullus** [springfall2008/batpred#1704]
**Describe the bug**
Octopus Go has a simple Peak and Off-Peak period. There is no financial or user benefit of exporting any battery until the end of the peak period prior to a recharge. Sometimes the plan creates 2 issues:

1. Export slots appear randomly in the middle of the peak period.
2. The last end of day export is not pushed out to the 00:00-00:30 slot but remains in the 23:30-00:00 slot. 

**Expected behaviour**
Export slots (if any) should all be at the end of the peak period to avoid running out of battery before the battery recharges in the off-peak period.

**Predbat version**

Main - 8.8.3(?)

**Environment details**

- GE AIO
- Standard HAOS installer

**Screenshots**

End of day export finishing early:

![image](https://github.com/user-attachments/assets/6477126f-5ae3-4ca1-9750-23ac95e0b2b8)

Random export slots in the day:

![image](https://github.com/user-attachments/assets/b7144a9f-5178-4ec2-8b6e-964d70d1f616)

**Log file**
Can you capture a log file from the time of the issue, debug mode is not normally required.
If you are not keeping the full logs then please enable this in appdaemon.yaml (see the installation instruction in the Predbat docs area for details on how to do this)

Debug Log File from 10:00 this morning - rname .txt back to .yaml

[predbat_debug_10_00_00.txt](https://github.com/user-attachments/files/18051636/predbat_debug_10_00_00.txt)

**Comment by springfall2008** [comment]
Great, the new debug method worked and I was able to reproduce your plan:

![image](https://github.com/user-attachments/assets/9f97fe00-ac97-46a7-88a0-1c142a7b5217)

**Comment by springfall2008** [comment]
So I think the behaviour is due to 'combine_export_slots' being True and thus the slots being too big

**Comment by springfall2008** [comment]
That doesn't fully fix it however, debugging

**Comment by springfall2008** [comment]
Fix released as v8.8.3

**Comment by nbullus** [comment]
Ok. Without it on the export at end of day is split and intermixes, I.e, charge, export, charge, export, charge , export, charge and so on. Better to do what it shows , ie charge, export, charge.Sent from my iPhoneOn 8 Dec 2024, at 15:27, Trefor Southwell ***@***.***> wrote:﻿
So I think the behaviour is due to 'combine_export_slots' being True and thus the slots being too big

**Comment by nbullus** [comment]
I've turned off 'Combine Charge Slots' and installed latest release, but my plan looks the same. I still have rogue exports appearing in the day and the last export of the day is still finishing 30 mins early.

![image](https://github.com/user-attachments/assets/fefd3a21-c48d-47cc-b58b-1930e03cdcfd)

![image](https://github.com/user-attachments/assets/3a401617-b80a-42af-9922-5c9e72854635)

**Comment by nbullus** [comment]
[predbat_debug_14_00_00.txt](https://github.com/user-attachments/files/18062616/predbat_debug_14_00_00.txt)

**Comment by mpartington** [comment]
Here's one from me. Discharge on charge slots is switched off, so shouldn't be doing this. I'd expect it to discharge without the charge cycles in-between with this set. Log to follow
![Screenshot_2024-12-09-16-34-51-719_io homeassistant companion android](https://github.com/user-attachments/assets/1322c48c-b9c1-499b-b08f-d2e90e71d844)

**Comment by mpartington** [comment]
I'm afraid I don't have the debug logs, but you can see how many random very short extort periods have been set today.
![Screenshot_2024-12-09-21-42-42-270_io homeassistant companion android~2](https://github.com/user-attachments/assets/3db9b1da-2986-4d39-a996-06c5a1c052d9)

**Comment by mpartington** [comment]
Another 2 min discharge (5 mins after I turned debug off). Now its planning split slots, rather than combine.

<img width="302" alt="image" src="https://github.com/user-attachments/assets/bb9fe98a-5fbe-4c18-942d-a3d1a30d0d2a">

[predbat_debug_22_10_00.txt](https://github.com/user-attachments/files/18068745/predbat_debug_22_10_00.txt)

**Comment by springfall2008** [comment]
Fix on 'main' for testing which appears to resolve Nigel's issue. Will re-test Matt's one now

**Comment by springfall2008** [comment]
Appears to fix Matt's case too

![image](https://github.com/user-attachments/assets/d6ed42be-282b-4498-8108-3e9d3ec8816a)

**Comment by mpartington** [comment]
Here are the files for the discharge in the charge slots (with setting off) 

[predbat_debug_20_35_00.txt](https://github.com/user-attachments/files/18085714/predbat_debug_20_35_00.txt)

[predbat_debug_20_40_00.txt](https://github.com/user-attachments/files/18085712/predbat_debug_20_40_00.txt)

**Comment by mpartington** [comment]
More random and very short export slots, incredibly annoying and burning unnecessary writes

<img width="812" alt="image" src="https://github.com/user-attachments/assets/1bf3c814-58c8-4b52-9fe6-4b34a4feea85" />
<img width="331" alt="image" src="https://github.com/user-attachments/assets/a5b87ca2-e932-4544-b20b-0ffc214467f1" />

Relevant Debug
[predbat_debug_16_30_00.txt](https://github.com/user-attachments/files/18099875/predbat_debug_16_30_00.txt)
[predbat_debug_16_25_00.txt](https://github.com/user-attachments/files/18099878/predbat_debug_16_25_00.txt)
[predbat_debug_17_00_00.txt](https://github.com/user-attachments/files/18099879/predbat_debug_17_00_00.txt)
[predbat_debug_16_55_00.txt](https://github.com/user-attachments/files/18099881/predbat_debug_16_55_00.txt)

[predbat1.log](https://github.com/user-attachments/files/18099897/predbat1.log)
[predbat.log](https://github.com/user-attachments/files/18099899/predbat.log)

**Comment by springfall2008** [comment]
Another version on 'main' for testing which may help

**Comment by mpartington** [comment]
> Another version on 'main' for testing which may help

Thank you, not much left to export tonight, will be running on fumes, but will leave debug on and feedback tomorrow

**Comment by mpartington** [comment]
Actually, might have one already
![Screenshot_2024-12-11-20-21-05-175_io homeassistant companion android](https://github.com/user-attachments/assets/3a57c9bb-e48d-4cb7-85a2-2a996b847b41)
[predbat_debug_20_20_00.txt](https://github.com/user-attachments/files/18101936/predbat_debug_20_20_00.txt)

**Comment by springfall2008** [comment]
So I understand why now.

The optimal export amount is less than 10 minutes worth which is the minimum export slot size, however if you do it at a time with higher predicted load then you end up exporting slightly less which then prevents you running out of battery. The slot moves around depending on the exact load prediction and what battery you might have left (which changes as you use it).

Not sure on the right solution, I could make the minimum export slot bigger e.g. 15 minutes or we could try a bigger metric_min_improvement_export but that is scaled with slot size also

**Comment by springfall2008** [comment]
The earlier slots are because those have higher load predictions meaning the really export less.

I could make the minimum slot size 5 minutes also, but its more register writes....

**Comment by springfall2008** [comment]
New release for testing:

https://github.com/springfall2008/batpred/releases/tag/v8.8.5

Lets hope its not broken stuff.

**Comment by mpartington** [comment]
No, definitely not suggesting going down to 5 mins slots. I'm on 1 write today so far and want to keep that streak 😉.

Absolutely no excess today, but fortunately it hasn't tried to random  discharge either. A larger export slot may make sense, but is there anything stopping combining a smaller and larger e.g. a 30min.+ 5min =35.

I don't see too much issue with any size slot, but ideally if it could be scheduled to end at the transition to cheap rate and work backwards from there for the start time, when all import & export slots are the same.

**Comment by nbullus** [comment]
v8.8.5 still doing random exports for me. Turned my automation back on to stop exports in daytime.

**Comment by mpartington** [comment]
> v8.8.5 still doing random exports for me. Turned my automation back on to stop exports in daytime.

Have you changed input_number.predbat_metric_min_improvement_export to 5?

Upload a debug file id you catch it, or leave debug turned on for a while. Just need to make sure you regularly clean it out, as it adds up quickly in terms of file size

**Comment by nbullus** [comment]
> v8.8.5 still doing random exports for me. Turned my automation back on to stop exports in daytime.

No. Just spotted v8.8.6 was released so will see how this version behaves as the min improvement export is now 5p by default

**Comment by mpartington** [comment]
You still need to manually change it, if you are using expert mode

**Comment by mpartington** [comment]
Latest version and recommended settings
![Screenshot_2024-12-14-12-32-03-785_io homeassistant companion android](https://github.com/user-attachments/assets/bad5fb22-b6ba-4d61-b404-10abac7aa6c5)
[Uploading predbat_debug_12_30_00.txt…]()

**Comment by PianSom** [comment]
Still having random import and export slots here on 8.8.9 - see pattern in blue ovals below. 

Also, green ovals show early export leaving a shortfall later in the evening. (I was away last night, so load would have been lower than forecast - so not that.)

Happy to supply debug file if helpful.
 
![IMG_7749](https://github.com/user-attachments/assets/dd888b9c-14ab-4eac-87ae-ca39f9979843)

**Comment by mpartington** [comment]
Are you using expert mode, if so have you set Metric Min Improvement Export to 5 as well?

**Comment by PianSom** [comment]
> Are you using expert mode, if so have you set Metric Min Improvement Export to 5 as well?

Yes, and yes

**Comment by PianSom** [comment]
I am on IOG (so cheap 11.30-5.30). At this time of year my battery is usually empty by the end of the evening. One very odd thing is that I can almost guarantee that every night I will have a short charge-then-discharge cycle in the period 11.30-midnight - for reasons which completely escape me.

**Comment by mpartington** [comment]
I'm afraid the return of the split discharge.  Also why is predbat changing my discharge power?
Afraid no debug yaml, but have attached normal logs

<img width="844" alt="image" src="https://github.com/user-attachments/assets/d38f27f7-dbd3-42da-9904-a9b62105d123" />
[predbat.log](https://github.com/user-attachments/files/18195252/predbat.log)

**Comment by mpartington** [comment]
Another poor plan, unfortunately lots of stop start already. I don't know why load is zero, but that should cause more export.

Before and now plans

![Screenshot_2024-12-23-20-52-00-172_io homeassistant companion android](https://github.com/user-attachments/assets/e59f76f9-f501-4583-92f9-fe62280b4dfc)
![Screenshot_2024-12-23-21-43-38-740_io homeassistant companion android](https://github.com/user-attachments/assets/4cbfeec3-26aa-47a3-b2ff-478c41e947e3)
![Screenshot_2024-12-23-21-46-14-322_io homeassistant companion android](https://github.com/user-attachments/assets/a79edb30-5c18-4033-9e16-d9942bd3f780)

[predbat_debug.yaml.txt](https://github.com/user-attachments/files/18233651/predbat_debug.yaml.txt)

**Comment by mpartington** [comment]
And another itteration
![Screenshot_2024-12-23-21-51-45-722_io homeassistant companion android](https://github.com/user-attachments/assets/d14e44c5-3a83-4c77-b1de-0f947de8cd8b)
[predbat_debug.yaml.txt](https://github.com/user-attachments/files/18233668/predbat_debug.yaml.txt)

**Comment by springfall2008** [comment]
I think its fixed on latest

![image](https://github.com/user-attachments/assets/2c4c6c54-6854-4fdd-b7cb-33f6383afb2c)

But you should probably set: metric_min_improvement_export to 5

**Comment by gcoan** [comment]
> I think its fixed on latest

Christmas Day 🎄🎅, don't you get time off Trefor ?

Happy Christmas

**Comment by mpartington** [comment]
Thanks Trefor, hope you had a great day yesterday (hopefully away from too much Predbat debugging)

The new version does seem better, only 1 break in discharge. I already have set metric_min_improvement_export to 5, but will increase the Battery rate max scaling discharge a bit more.

Just when you have time, do you know why Predbat goes into hold exporting at a SOC of 6%?  This is repeatable and probably a good idea if Predbat is managing discharge, as prevents using the peak if it finishes a few minutes early.

However I sometimes want to force discharge until it stops discharging to allow the battery to settle and perform its OCV checks in an idle state.  I've noticed it still stops at 6% even using the manual overrides.

The screenshot below is last night, with no overrides set, to show the behaviour

<img width="823" alt="image" src="https://github.com/user-attachments/assets/584b0ed6-b118-4270-8b48-27ab08a6af64" />
[predbat_debug_23_25_00.yaml.txt](https://github.com/user-attachments/files/18251106/predbat_debug_23_25_00.yaml.txt)
[predbat_debug_23_30_00.yaml.txt](https://github.com/user-attachments/files/18251107/predbat_debug_23_30_00.yaml.txt)

**Comment by springfall2008** [comment]
I'd guess its because you have best_soc_keep as 0.5 and hence going any lower would violate that? Maybe turn that to 0 to disable keep?

**Comment by mpartington** [comment]
I did try that on the first chart above (where I did manual export slots). It unfortunately didn't make any difference 
![Screenshot_2024-12-26-10-55-20-212_io homeassistant companion android](https://github.com/user-attachments/assets/78bfd4ad-1241-4bbe-ad6a-dfc1788046f9)

**Comment by springfall2008** [comment]
I've made a new release with some tweaks that after reducing soc_keep to 0.4 gets me to this based on your YAML:

![image](https://github.com/user-attachments/assets/73c5e80f-6744-4465-8352-a4885eba6d10)

**Comment by mpartington** [comment]
Seems to have fixed my export to empty, in fact I ran out with 1 min to go.  I guess just need feedback that it does still respect a higher soc keep.  Additionally the discharge was in a single slot, which minimises register writes  and late as possible 👍 (although I see some beta firmware has already been released for AIO, and gen 2 & 3 to address this. Just waiting on gen 1)

**Comment by mpartington** [comment]
This plan doesn't seem correct. 1hr left and only 20 % battery. That's a max of 1.7kWh, with a discharge rate of 3000W, that's only 30 mins

![Screenshot_2024-12-27-22-39-23-253_io homeassistant companion android](https://github.com/user-attachments/assets/22d1430a-f616-4ee3-a57e-3d3ac5ebcf72)

[predbat_debug_22_35_00.yaml.txt](https://github.com/user-attachments/files/18264134/predbat_debug_22_35_00.yaml.txt)

[predbat.log](https://github.com/user-attachments/files/18264139/predbat.log)

**Comment by springfall2008** [comment]
You can model the discharge rate using the discharge curve?

**Comment by chalfontchubby** [comment]
Revisiting this during triage - the thread ended with confirmation the fix worked ("Seems to have fixed my export to empty") and there's been no further activity or recurrence report since. Closing as resolved; please reopen if this comes back.
